source/vibe/core/sync.d

```diff
@@ -819,3 +819,469 @@ private struct TaskConditionImpl(bool INTERRUPTIBLE, LOCKABLE) {
+     *      `false` otherwise.
+     */
+    @trusted bool tryLock(LockingIntent INTENT)()
+    {
+        //Try to lock this mutex without waiting for the underlying
+        //TaskMutex - fail if it is already blocked.
+        return tryLock!(INTENT,false)();
+    }
+    
+    /** Acquire the lock for the given intent; yield and suspend until the lock has been acquired. */
+    @trusted void lock(LockingIntent INTENT)()
+    {
+        //Prepare a waiting action before the first
+        //`tryLock()` call in order to avoid a race
+        //condition that could lead to the queue notification
+        //not being fired.
+        auto count = emitCount!INTENT;
+        atomicOp!"+="(queueCounter!INTENT,1);
+        scope(exit)
+            atomicOp!"-="(queueCounter!INTENT,1);
+        
+        //Try to lock the mutex
+        auto locked = tryLock!(INTENT,true)();
+        if (locked)
+            return;
+        
+        //Retry until we successfully acquired the lock
+        while(!locked)
+        {
+            version(RWMutexPrint)
+                printInfo!("wait",INTENT)();
+            
+            count  = wait!INTENT(count);
+            locked = tryLock!(INTENT,true)();
+        }
+    }
+    
+    /** Unlock the mutex after a successful acquisition. */
+    @trusted void unlock(LockingIntent INTENT)()
+    {
+        version(RWMutexPrint)
+            printInfo!("unlock",INTENT)();
+        
+        debug assert(activeCounter!INTENT > 0);
+
+        synchronized(m_counterMutex)
+        {
+            //Decrement the counter of active lock holders.
+            //If the counter hits zero, notify waiting Tasks
+            activeCounter!INTENT -= 1;
+            if (activeCounter!INTENT == 0)
+            {
+                version(RWMutexPrint)
+                    printInfo!("notify",INTENT)();
+                
+                notify!INTENT();
+            }
+        }
+    }
+}
+
+/** A ReadWriteMutex implementation for fibers.
+ *
+ *  This mutex can be used in exchange for a $(D core.sync.mutex.ReadWriteMutex),
+ *  but does not block the event loop in contention situations. The `reader` and `writer`
+ *  members are used for locking. Locking the `reader` mutex allows access to multiple 
+ *  readers at once, while the `writer` mutex only allows a single writer to lock it at
+ *  any given time. Locks on `reader` and `writer` are mutually exclusive (i.e. whenever a 
+ *  writer is active, no readers can be active at the same time, and vice versa).
+ * 
+ *  Notice:
+ *      Mutexes implemented by this class cannot be interrupted
+ *      using $(D vibe.core.task.Task.interrupt()). The corresponding
+ *      InterruptException will be deferred until the next blocking
+ *      operation yields the event loop.
+ * 
+ *      Use $(D InterruptibleTaskReadWriteMutex) as an alternative that can be
+ *      interrupted.
+ * 
+ *  cf. $(D core.sync.mutex.ReadWriteMutex)
+ */
+class TaskReadWriteMutex
```

Comment:
As I already noted, I had similar issues creating a global shared `TaskMutex` instance in the same fashion as my previous tests for the read-write version; so I decided to just revert to the basic non-template version for the two mutex classes.

If the missing shared overloads are an issue, I'd propose to solve that in a separate PR for all mutex classes in one go.